davinci-resolve: 18.5.1 -> 18.6.2; add studio variant; various improvements

amarshall opened the pull request:
## Description of changes

See individual commit messages for details and caveats.

## Things done

Tested both free and studio. Tested studio activation. Tested that app starts and can load existing project and various functionality. Using Nvidia GPU.

The Studio download is a bit flaky and sometimes fails with

> curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)

but works eventually.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

amarshall committed:
davinci-resolve: 18.5.1 -> 18.6

Without libXt, shows popup on launch:

> Unable to load plugin: USD.plugin
>
> libXt.so.6: cannot open shared object file: No such file or directory

ultimately launches okay without it, but some feature would probably be
broken.

amarshall committed:
davinci-resolve: passthru underlying drv

Otherwise getting it is…non-trivial.

amarshall committed:
davinci-resolve: fail fast if filename is incorrect

`appimage-run` fails silently if the file does not exist. This made it
easy to have mismatches that lead to non-obvious build errors.

amarshall committed:
davinci-resolve: add studioVariant

Resolve wants to write the license to `./.license` in the drv store
path. This obviously fails as it is read-only. Conveniently this is a
Bubblewrap-based FHS env, so we can bind-mount a reasonable dir from
$HOME to this path instead. Unfortunately, the dir needs to exist before
launching, so it must be manually created presently e.g.

    mkdir -p ~/.local/share/DaVinciResolve/license

At least the error is fairly obvious:

> bwrap: Can't find source path /home/user/.local/share/DaVinciResolve/license: No such file or directory

Without making the license writable, license activation will always give
a generic error—the actual problem was found with strace.

amarshall committed:
davinci-resolve: dynamically get downloadid

The downloads.json also includes metadata for older versions, so this
should be durable unless upstream breaks things. Some older downloads
appear to not include the “Update” suffix in the name, so match with or
without it.

amarshall committed:
davinci-resolve: fix missing version in meta

See https://github.com/NixOS/nixpkgs/issues/253232

stelcodes commented on pkgs/applications/video/davinci-resolve/default.nix:
```diff
@@ -47,19 +50,24 @@ let
-          SITEURL = "https://www.blackmagicdesign.com/api/register/us/download/${DOWNLOADID}";
+          DOWNLOADSURL = "https://www.blackmagicdesign.com/api/support/us/downloads.json";
+          SITEURL = "https://www.blackmagicdesign.com/api/register/us/download";
+          PRODUCT = "DaVinci Resolve${if studioVariant then " Studio" else ""}";
```
Same here, `lib.optionalString`

stelcodes commented on pkgs/applications/video/davinci-resolve/default.nix:
```diff
@@ -22,13 +22,16 @@
     stdenv.mkDerivation rec {
-      pname = "davinci-resolve";
-      version = "18.5.1";
+      pname = "davinci-resolve${if studioVariant then "-studio" else ""}";
```
Better to use `lib.optionalString` here

stelcodes reviewed (commented):
A couple comments

amarshall committed:
davinci-resolve: 18.6 -> 18.6.2

amarshall committed:
davinci-resolve: 18.6.2 -> 18.6.3

amarshall committed:
davinci-resolve: fix runtime error

See https://github.com/NixOS/nixpkgs/issues/265476

This is a hack that effectively undoes
https://github.com/NixOS/nixpkgs/pull/263201 for just this derivation.
Really we should probably be doing a ton of autoPatchelf or something to
patch in our libs. Alas, this is “good enough” to un-break for now, I
think.

xvrqt commented:
I bought a license to try studio version;  it builds fine, but upon running it complains that it can't copy the license file into the read-only `/nix/store` path (makes sense). 

It seems like a recreation of an older issue: https://github.com/NixOS/nixpkgs/issues/94032#issuecomment-921918602 where davinci-resolve reverts to its `/nix/store` location no matter where it is run from. This was seeming solved by wrapping it in a FHS location built for davinci using https://github.com/NixOS/nixpkgs/blob/bcd452a5544131d486eac0493002d2341b3af022/pkgs/build-support/build-fhs-userenv-bubblewrap/default.nix. 

The issue was brought up again in https://github.com/NixOS/nixpkgs/pull/152113#issuecomment-1075642604 where the work around was to use `extraBwrapArgs` to bind the license file to a writeable location in the FHS. However this is failing for me :[ 

I am installing it via a standalone home-manager installation, and pulling nixpkgs-unstable via a flake.
Installing it as a `systemPackages` on my NixOS configuration produces the same error.

`bwrap: Can't create file at /nix/store/kxy0cl1p0f9nr6gf46gzc1jx42f48kzw-davinci-resolve-studio-18.6.3/.license: Is a directory`

I'm pretty green to NixOS so I'm trying to understand `buildFHSEnv` better to see why davinci isn't running from there, but would appreciated if someone more knowledgeable would take a look.

xvrqt commented:
Seems related: https://github.com/NixOS/nixpkgs/issues/239017

Yet, we're clearly using `--bind` in the `extraBwrapArgs` so idk what's up :[

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/davinci-resolve-studio-install-issues/37699/5
